taskcat/amiupdater.py

```diff
@@ -0,0 +1,409 @@
+        self._objs.append(self)
+
+        # Generate RegionalCodename filters based on what's in the template. 
+        self._generate_regional_codenames()
+
+    def _generate_regional_codenames(self):
+        self.filters = None
+        self._region_list = list()
+        _ec2_regions = AMIUpdater.client_factory.get('ec2', 'us-east-1').describe_regions()['Regions']
+        for _ec2r in _ec2_regions:
+            self._region_list.append(_ec2r['RegionName'])
+        if self.all_regions:
+            for region in self._region_list:
+                self._regions.add(region)
+        else:
+            # Use the regions that are in Mappings/AWSAMIRegionMap
+            for region in self._mapping_root.keys():
+                if region not in self._region_list:
+                    raise AMIUpdaterException("Template: [{}] Region: [{}] is not a valid region".format(self._filename, region))
+                self._regions.add(region)
+                
+        self.codename = None
+        # Looking for Mappings/AWSAMIRegionMap
+        if not self._mapping_root:
+            return None
+        # This is where we know the instantation is good (we've passed sanity checks).
+        # Appending the object so it can be referenced later.
+        self._objs.append(self)
+        if self.filter_metadata:
+            for k in self.filter_metadata.keys():
+                for region in self._regions:
+                    if region == 'AMI':
+                        continue
+                    RegionalCodename(cn=k, region=region, filters=self.filter_metadata[k])
+        else:
+            # For reach region
+            for region in self._regions:
+                # 'AMI' is in the list, skip it.
+                if region == 'AMI':
+                    continue
+                # Region Name, Latest AMI Name in Template
+                # - We instantiate them in the RegionalCodename class
+                #   because it allows us to access the attributes as an object.
+                #   It also generates the Filters needed in each API call.
+                #   - This is done in the Codenames class, so check that out.
+                for k in self._mapping_root[region].keys():
+                    RegionalCodename(cn=k, region=region)
+
+    def _determine_regions(self):
+        self._region_list = list()
+        _ec2_regions = AMIUpdater.client_factory.get('ec2', 'us-east-1').describe_regions()['Regions']
+        for _ec2r in _ec2_regions:
+            self._region_list.append(_ec2r['RegionName'])
+        if self.all_regions:
+            for region in self._region_list:
+                self._regions.add(region)
+        else:
+            # Use the regions that are in Mappings/AWSAMIRegionMap
+            for region in self._mapping_root.keys():
+                if region not in self._region_list:
+                    raise AMIUpdaterException("Template: [{}] Region: [{}] is not a valid region".format(self._filename, region))
```

Comment:
this breaks if the legacy `AMI` key is present

amiupdater ~/quickstart-amazon-eks/templates/amazon-eks-nodegroup.template.yaml
Traceback (most recent call last):
  File "/Users/jmmccon/taskcat/venv-pc/bin/amiupdater", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/jmmccon/taskcat/bin/amiupdater", line 38, in <module>
    a.update_amis()
  File "/Users/jmmccon/taskcat/taskcat/amiupdater.py", line 390, in update_amis
    TemplateObject(template_file)
  File "/Users/jmmccon/taskcat/taskcat/amiupdater.py", line 252, in __init__
    self._determine_regions()
  File "/Users/jmmccon/taskcat/taskcat/amiupdater.py", line 320, in _determine_regions
    raise AMIUpdaterException("Template: [{}] Region: [{}] is not a valid region".format(self._filename, region))
taskcat.amiupdater.AMIUpdaterException: Template: [/Users/jmmccon/quickstart-amazon-eks/templates/amazon-eks-nodegroup.template.yaml] Region: [AMI] is not a valid region